tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,246 @@
+    await gossipSub.switch.stop()
+
+  # Test ensuring that the number of subscriptions does not exceed the limit set in the GossipSub parameters
+  asyncTest "subscription limit test":
+    let topicCount = 15
+    let gossipSubParams = 10
+    let topicNames = toSeq(mapIt(0 .. topicCount - 1, "topic" & $it))
+
+    let (gossipSub, conns) = setupGossipSub(topicNames, 0)
+
+    gossipSub.topicsHigh = gossipSubParams
+
+    for topic in topicNames:
+      if gossipSub.topics.len < gossipSub.topicsHigh:
+        gossipSub.subscribe(
+          topic,
+          proc(topic: string, data: seq[byte]): Future[void] {.async.} =
+            discard
+          ,
+        )
+      else:
+        check gossipSub.topics.len == gossipSub.topicsHigh
+
+    check gossipSub.topics.len <= gossipSub.topicsHigh
+    check gossipSub.topics.len == gossipSub.topicsHigh
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test for verifying peers joining a topic using `JOIN(topic)`
+  asyncTest "handle JOIN topic and mesh is updated":
+    let topic = "test-join-topic"
+
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    commonSubscribe(@[gossipSub], topic, voidTopicHandler)
+
+    check gossipSub.mesh[topic].len == 5
+    check gossipSub.topics.contains(topic)
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test the behavior when multiple peers join and leave a topic simultaneously.
+  asyncTest "multiple peers join and leave topic simultaneously":
+    let
+      numberOfNodes = 6
+      topic = "foobar"
+      nodes = generateNodes(numberOfNodes, gossip = true)
+      nodesFut = await allFinished(nodes.mapIt(it.switch.start()))
+
+    await subscribeNodes(nodes)
+    for node in nodes:
+      node.subscribe(topic, voidTopicHandler)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = GossipSub(nodes[i])
+      check currentGossip.gossipsub.hasKey(topic)
+
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = GossipSub(nodes[i])
+
+    for x in 0 ..< numberOfNodes:
+      for y in 0 ..< numberOfNodes:
+        if x != y:
+          await waitSub(nodes[x], nodes[y], topic)
+
+    await sleepAsync(2 * DURATION_TIMEOUT)
+
+    let expectedNumberOfPeers = numberOfNodes - 1
+    for i in 0 ..< numberOfNodes:
+      let currentGossip = GossipSub(nodes[i])
+      check:
+        currentGossip.gossipsub[topic].len == expectedNumberOfPeers
+        currentGossip.mesh[topic].len == expectedNumberOfPeers
+        currentGossip.fanout.len == 0
+
+    let firstNodeGossip = GossipSub(nodes[0])
+    let peersToUnsubscribe = firstNodeGossip.mesh[topic].toSeq()[0 .. 2]
+    let peerIdsToUnsubscribe = peersToUnsubscribe.mapIt(it.peerId)
+    for node in nodes:
+      for peerId in peerIdsToUnsubscribe:
+        node.unsubscribe(topic, voidTopicHandler)
+        node.peers.del(peerId)
```

Comment:
You can't call `unsubscribe` from a `PubSubPeer`:
- `PubSubPeer` is a node's representation of each of its peers. Because it's just a representation/handler it can't call methods such as `unsubscribe` on the node's behalf.
- `unsubscribe` can only be called on `PubSub` instances, or its children. E.g.: `FloodSub`, `GossipSub`.

Due to reasons above, to `unsubscribe` a node you need to use the node itself:
```suggestion
    let 
      firstNodeGossip = GossipSub(nodes[0])
      peersToUnsubscribe = nodes[1 ..< 3]
    for peer in peersToUnsubscribe:
        peer.unsubscribe(topic, voidTopicHandler)
```